# Benchmark of software tools for prokaryotic chromosomal interaction domain identification

**Datasets used:**
* *Caulobacter crescentus* (Le *et al.*, *Science*, 2013)
* *Bacillus subtilis* (Wang *et al.*, *Genes Dev.*, 2015; Marbouty *et al.*, *Mol. Cell*, 2015)
* *Escherichia coli* (Lioy *et al.*, *Cell*, 2018)
* *Mycoplasma pneumoniae* (Trussart *et al.*, *Nat. Comm.*, 2017)
* *Sulfolobus acidocaldarius* (Takemata *et al.*, *Cell*, 2019)

**Compared domain-calling tools:**
* [Armatus](https://github.com/kingsfordgroup/armatus/) (Filippova *et al.*, *Algorithms Mol Biol.*, 2014)
* [Arrowhead](https://github.com/aidenlab/juicer/wiki/Arrowhead) (Durand *et al.*, *Cell Systems*, 2016)
* [ClusterTAD](https://github.com/BDM-Lab/ClusterTAD) (Oluwadare *et al.*, *BMC Bioinformatics*, 2017)
* [Directionality Index]() (Dixon *et al.*, *Nature*, 2012; Le *et al.*, *Science*, 2013)
* [HiCExplorer](https://github.com/deeptools/HiCExplorer) (Ramirez *et al.*, *Nat. Comm.*, 2018)
* [HiCseg](https://CRAN.R-project.org/package=HiCseg) (Levy-Leduc *et al.*, *Bioinformatics*, 2014)
* [Lavaburst](https://github.com/nvictus/lavaburst)
* [SpectralTAD](https://github.com/dozmorovlab/SpectralTAD) (Cresswell *et al.*, *bioRxiv*, 2019)
* [TADbit](https://github.com/3DGenomes/TADbit) (Serra *et al.*, *PLoS Comput. Biol.*, 2017)
* [TADtool](https://github.com/vaquerizaslab/tadtool) (Kruse *et al.*, *Bioinformatics*, 2016)
* [TADtree](http://compbio.cs.brown.edu/projects/tadtree/) (Weinreb *et al.*, *Bioinformatics*, 2015)
* [TopDom](https://github.com/HenrikBengtsson/TopDom) (Shin *et al.*, *Nucleic Acids Res.*, 2016)

**Available code and data:**
* Hi-C and RNA-seq data processing pipelines
* Scripts to run domain callers used for benchmarking
* Annotated domain positions used in the analysis
* Scripts to analyse and validate the annotated domains
* Scripts to digest genomes and simulate Hi-C read pairs
